crypto/hashtable/hashtable.c

```diff
@@ -0,0 +1,691 @@
+                                    &new->md->neighborhood_ptr_to_free);
+    if (new->md->neighborhoods == NULL)
+        goto err;
+    new->md->neighborhood_mask = new->wpd.neighborhood_len - 1;
+
+    new->lock = ossl_rcu_lock_new(1);
+    if (new->lock == NULL)
+        goto err;
+
+    if (new->config.ht_hash_fn == NULL)
+        new->config.ht_hash_fn = fnv1a_hash;
+
+    return new;
+
+err:
+    CRYPTO_THREAD_lock_free(new->atomic_lock);
+    ossl_rcu_lock_free(new->lock);
+    OPENSSL_free(new->md->neighborhood_ptr_to_free);
+    OPENSSL_free(new->md);
+    OPENSSL_free(new);
+    return NULL;
+}
+
+void ossl_ht_read_lock(HT *htable)
+{
+    ossl_rcu_read_lock(htable->lock);
+}
+
+void ossl_ht_read_unlock(HT *htable)
+{
+    ossl_rcu_read_unlock(htable->lock);
+}
+
+void ossl_ht_write_lock(HT *htable)
+{
+    ossl_rcu_write_lock(htable->lock);
+    htable->wpd.need_sync = 0;
+}
+
+void ossl_ht_write_unlock(HT *htable)
+{
+    int need_sync = htable->wpd.need_sync;
+
+    htable->wpd.need_sync = 0;
+    ossl_rcu_write_unlock(htable->lock);
+    if (need_sync)
+        ossl_synchronize_rcu(htable->lock);
+}
+
+static void free_oldmd(void *arg)
+{
+    struct ht_mutable_data_st *oldmd = arg;
+    size_t i, j;
+    size_t neighborhood_len = (size_t)oldmd->neighborhood_mask + 1;
+    struct ht_internal_value_st *v;
+
+    for (i = 0; i < neighborhood_len; i++) {
+        PREFETCH_NEIGHBORHOOD(oldmd->neighborhoods[i + 1]);
+        for (j = 0; j < NEIGHBORHOOD_LEN; j++) {
+            if (oldmd->neighborhoods[i].entries[j].value != NULL) {
+                v = oldmd->neighborhoods[i].entries[j].value;
+                v->ht->config.ht_free_fn((HT_VALUE *)v);
+                free_value(v);
+            }
+        }
+    }
+
+    OPENSSL_free(oldmd->neighborhood_ptr_to_free);
+    OPENSSL_free(oldmd);
+}
+
+static int ossl_ht_flush_internal(HT *h, int replace)
+{
+    struct ht_mutable_data_st *newmd = NULL;
+    struct ht_mutable_data_st *oldmd = NULL;
+
+    newmd = OPENSSL_zalloc(sizeof(*newmd));
+    if (newmd == NULL)
+        return 0;
+
+    if (replace == 1) {
+        newmd->neighborhoods = alloc_new_neighborhood_list(DEFAULT_NEIGH_LEN,
+                                                           &newmd->neighborhood_ptr_to_free);
+        if (newmd->neighborhoods == NULL) {
+            OPENSSL_free(newmd);
+            return 0;
+        }
+    }
+
+    newmd->neighborhood_mask = DEFAULT_NEIGH_LEN - 1;
+
+    /* Swap the old and new mutable data sets */
+    oldmd = h->md;
+    ossl_rcu_assign_ptr(&h->md, &newmd);
+
+    /* Set the number of entries to 0 */
+    h->wpd.value_count = 0;
+    h->wpd.neighborhood_len = DEFAULT_NEIGH_LEN;
+
+    ossl_rcu_call(h->lock, free_oldmd, oldmd);
+    h->wpd.need_sync = 1;
+    return 1;
+}
+
+int ossl_ht_flush(HT *h)
+{
+    return ossl_ht_flush_internal(h, 1);
+}
+
+void ossl_ht_free(HT *h)
+{
+    if (h == NULL)
+        return;
+
+    ossl_ht_write_lock(h);
+    ossl_ht_flush_internal(h, 1);
```

Comment:
yeah, I should actually refactor this.  For reference however, the h->md->neighborhoods pointer is freed via the aliased neighborhood_ptr_to_free pointer below.  This is done because of OPENSSL_aligned_alloc call.  In the event an OS based aligned alloc call doesn't exist, it may be that the pointer to free an allocated block differs from the aligned pointer.  The details of how that works are in the aligned_alloc docs